File: backend/src/mappers/assignment.ts
import type { Assignment } from 'vulms-sdk';
import type { AssignmentDto, AssignmentSummaryDto } from '../types/assignments.js';

export function toAssignmentDto(a: Assignment): AssignmentDto {
  return {
    courseCode: a.courseCode,
    courseTitle: a.courseTitle,
    title: a.title,
    lesson: a.lesson,
    dueDate: a.dueDate instanceof Date ? a.dueDate.toISOString() : typeof a.dueDate === 'string' ? a.dueDate : undefined,
    status: a.status,
    totalMarks: a.totalMarks,
    obtainedMarks: a.obtainedMarks,
  };
}

export function toAssignmentSummaryDto(summary: {
  total: number;
  submitted: number;
  pending: number;
  missed: number;
  resultDeclared: number;
}): AssignmentSummaryDto {
  return {
    total: summary.total,
    submitted: summary.submitted,
    pending: summary.pending,
    missed: summary.missed,
    resultDeclared: summary.resultDeclared,
  };
}

File: vulms_dart/lib/src/models/assignment.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'assignment.freezed.dart';
part 'assignment.g.dart';

/// Status of an assignment.
enum AssignmentStatus {
  pending,
  submitted,
  attempted,
  missed,
  resultDeclared,
}

/// Represents a VULMS assignment.
@freezed
abstract class Assignment with _$Assignment {
  const factory Assignment({
    required String courseCode,
    required String courseTitle,
    required String title,
    String? lesson,
    DateTime? dueDate,
    double? totalMarks,
    @Default(AssignmentStatus.pending) AssignmentStatus status,
    DateTime? submitDate,
    String? fileSize,
    double? obtainedMarks,
  }) = _Assignment;

  factory Assignment.fromJson(Map<String, dynamic> json) =>
      _$AssignmentFromJson(json);
}
